plugin/src/main/java/org/opensearch/ml/action/models/DeleteModelTransportAction.java

```diff
@@ -333,6 +478,59 @@ private Boolean isModelNotDeployed(MLModelState mlModelState) {
             && !mlModelState.equals(MLModelState.PARTIALLY_DEPLOYED);
     }
 
+    private <T> List<String> findDependentPipelines(
+        List<T> pipelineConfigurations,
+        String candidateModelId,
+        Function<T, Map<String, Object>> getConfigFunction,
+        Function<T, String> getIdFunction
+    ) {
+        List<String> dependentPipelineConfigurations = new ArrayList<>();
+        for (T pipelineConfiguration : pipelineConfigurations) {
+            Map<String, Object> config = getConfigFunction.apply(pipelineConfiguration);
+            if (searchThroughConfig(config, candidateModelId, "")) {
+                dependentPipelineConfigurations.add(getIdFunction.apply(pipelineConfiguration));
+            }
+        }
+        return dependentPipelineConfigurations;
+    }
+
+    // This method is to go through the pipeline configs and only when the key is model id and value is
+    // 1. String and equal to candidate id 2. A list of String containing candidate id We will return True. Otherwise False
+    private Boolean searchThroughConfig(Object searchCandidate, String candidateId, String targetModelKey) {
+        // Use a stack to store the elements to be processed
+        Deque<Pair<String, Object>> stack = new ArrayDeque<>();
+        stack.push(Pair.of(targetModelKey, searchCandidate));
+
+        while (!stack.isEmpty()) {
+            // Pop an item from the stack
+            Pair<String, Object> current = stack.pop();
+            String currentKey = current.getLeft();
+            Object currentCandidate = current.getRight();
+
+            if (currentCandidate instanceof String) {
+                // Check for a match
+                if (Objects.equals(currentKey, PIPELINE_TARGET_MODEL_KEY) && Objects.equals(candidateId, currentCandidate)) {
+                    return true;
+                }
+            } else if (currentCandidate instanceof List<?>) {
+                // Push all elements in the list onto the stack
+                for (Object v : (List<?>) currentCandidate) {
+                    stack.push(Pair.of(currentKey, v));
+                }
+            } else if (currentCandidate instanceof Map<?, ?>) {
+                // Push all values in the map onto the stack
+                for (Map.Entry<?, ?> entry : ((Map<?, ?>) currentCandidate).entrySet()) {
+                    String key = (String) entry.getKey();
```

Comment:
entry.getKey().toString() is safer

Reply:
It's fine because we can assume the keys are all string, and if in edge case the key is null, (String) entry.getKey() won't throw exception but entry.getKey().toString() will, so let's keep it as is.